zerver/lib/test_classes.py

```diff
@@ -167,6 +184,35 @@ def run(self, result: Optional[TestResult]=None) -> Optional[TestResult]:  # noc
     DEFAULT_SUBDOMAIN = "zulip"
     TOKENIZED_NOREPLY_REGEX = settings.TOKENIZED_NOREPLY_EMAIL_ADDRESS.format(token="[a-z0-9_]{24}")
 
+    @contextmanager
+    def assertNoLogs(self, test_case,logger_name, level, no_logs, exc_type, exc_value,tb):
+        LOGGING_FORMAT = "%(levelname)s:%(name)s:%(message)s"
+        formatter = logging.Formatter(LOGGING_FORMAT)
+        handler = _CapturingHandler()
+        handler.setLevel(level)
+        handler.setFormatter(formatter)
+        self.watcher = handler.watcher
+        self.old_handlers = logger.handlers[:]
+        self.old_level = logger.level
+        self.old_propagate = logger.propagate
+        logger.handlers = [handler]
+        logger.setLevel(level)
+        logger.propagate = False
+        if self.no_logs:
+            yield
+        yield handler.watcher
+        self.logger.handlers = self.old_handlers
+        self.logger.propagate = self.old_propagate
+        self.logger.setLevel(self.old_level)
+        if exc_type is not None:
+            return False
+        if self.no_logs:
+            if len(self.watcher.records) > 0:
+                self._raiseFailure("Unexpected logs found: {!r}".format(self.watcher.output))
+        else:
+            if len(self.watcher.records) == 0:
```

Comment:
This part has nothing to do with `assertNoLogs` as I think `no_logs` should always be true. So we need only that case. You have included `assertLogs` stuff here which is not required.

Reply:
Please have a look at new solution posted.